presto-delta/src/main/java/com/facebook/presto/delta/DeltaTableName.java

```diff
@@ -0,0 +1,144 @@
+                    "(?:@v(?<snapshotId>[0-9]+))?" +
+                    "(?:@t(?<timestamp>[0-9.\\-: ]+))?");
+
+    private static final DateTimeFormatter TIMESTAMP_PARSER =
+            new DateTimeFormatterBuilder().appendPattern("yyyy[-MM[-dd[ HH[:mm[:ss]]]]]")
+                    .parseDefaulting(ChronoField.MONTH_OF_YEAR, 1)
+                    .parseDefaulting(ChronoField.DAY_OF_MONTH, 1)
+                    .parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
+                    .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+                    .parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+                    .toFormatter()
+                    .withZone(ZoneId.of("UTC"));
+
+    private final String tableNameOrPath;
+    private final Optional<Long> snapshotId;
+    private final Optional<Long> timestampMillisUtc;
+
+    public DeltaTableName(String tableNameOrPath, Optional<Long> snapshotId, Optional<Long> timestampMillisUtc)
+    {
+        this.tableNameOrPath = requireNonNull(tableNameOrPath, "tableNameOrPath is null");
+        this.snapshotId = requireNonNull(snapshotId, "snapshotId is null");
+        this.timestampMillisUtc = requireNonNull(timestampMillisUtc, "timestampMillisUtc is null");
+    }
+
+    public String getTableNameOrPath()
+    {
+        return tableNameOrPath;
+    }
+
+    public Optional<Long> getSnapshotId()
+    {
+        return snapshotId;
+    }
+
+    public Optional<Long> getTimestampMillisUtc()
+    {
+        return timestampMillisUtc;
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder builder = new StringBuilder()
+                .append("Table[" + tableNameOrPath + "]");
+        snapshotId.map(id -> builder.append("@v" + id));
+        timestampMillisUtc.map(ts -> builder.append("@t" + new Timestamp(ts)));
+        return builder.toString();
+    }
+
+    public static DeltaTableName from(String tableName)
+    {
+        Matcher match = TABLE_PATTERN.matcher(tableName);
+        if (!match.matches()) {
+            throw new PrestoException(NOT_SUPPORTED, "Invalid Delta table name: " + tableName +
+                    ", Expected table name form 'tableName[@v<snapshotId>][@t<snapshotAsOfTimestamp>]'. " +
+                    "The table can have either a particular snapshot identifier or a timestamp of the snapshot. " +
+                    "If timestamp is given the latest snapshot of the table that was generated at or before the given timestamp is read");
+        }
+
+        String tableNameOrPath = match.group("table");
+        String snapshotValue = match.group("snapshotId");
+        Optional<Long> snapshot = Optional.empty();
+        if (snapshotValue != null) {
+            snapshot = Optional.of(Long.parseLong(snapshotValue));
+        }
+
+        Optional<Long> timestampMillisUtc = Optional.empty();
+        String timestampValue = match.group("timestamp");
```

Comment:
could we define `table`, `snapshotId`, `timestamp` as constant? or static import them?